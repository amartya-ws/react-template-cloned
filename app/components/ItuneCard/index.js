/**
 *
 * ItuneCard
 *
 */
import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { useHistory } from 'react-router-dom';
import { Card } from 'antd';

const StyledCard = styled(Card)`
  && {
    width: 18rem;
    min-height: 21rem;
  }
`;

const StyledImg = styled.img`
  && {
    max-height: 14.5rem;
    object-fit: cover;
  }
`;

const StyledAudio = styled.audio`
  && {
    width: 100%;
    height: 2.5rem;
    margin-top: 1rem;
  }
`;

export function ItuneCard({
  artistName,
  trackName,
  artworkUrl100,
  previewUrl,
  currentSongId,
  setCurrentSongId,
  trackId
}) {
  const { Meta } = Card;
  const audioRef = useRef(null);
  const history = useHistory();

  useEffect(() => {
    if (currentSongId !== trackId) {
      audioRef.current.pause();
    }
  }, [currentSongId]);

  const playHandler = () => {
    setCurrentSongId(trackId);
  };

  const clickHandler = () => {
    history.push(`/song/${trackId}`);
  };

  return (
    <div data-testid="itune-card" onClick={clickHandler}>
      <StyledCard cover={<StyledImg alt={'song cover'} src={artworkUrl100} data-testid="song-image" />}>
        <Meta title={trackName ?? 'not found'} description={artistName ?? 'not found'} data-testid="song-detail" />
        <StyledAudio
          data-testid="preview-audio"
          controls
          src={previewUrl}
          ref={audioRef}
          onPlay={playHandler}
          controlsList="nodownload noplaybackrate"
        >
          Audio preview
        </StyledAudio>
      </StyledCard>
    </div>
  );
}

ItuneCard.propTypes = {
  artistName: PropTypes.string,
  trackName: PropTypes.string,
  artworkUrl100: PropTypes.string,
  previewUrl: PropTypes.string,
  currentSongId: PropTypes.number,
  setCurrentSongId: PropTypes.func,
  trackId: PropTypes.number
};

export default ItuneCard;
