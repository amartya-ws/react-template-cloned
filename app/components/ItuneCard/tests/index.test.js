/**
 *
 * Tests for ItuneCard
 *
 */

import React from 'react';
import { renderWithIntl } from '@utils/testUtils';
import ItuneCard from '../index';

describe('<ItuneCard />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<ItuneCard />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 ItuneCard component', () => {
    const { getAllByTestId } = renderWithIntl(<ItuneCard />);
    expect(getAllByTestId('itune-card').length).toBe(1);
  });

  it('should render itune song details inside the card', () => {
    const data = {
      artistName: 'eminem',
      trackName: 'Till I Collapse (feat. Nate Dogg)',
      url: 'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/b4/13/e9/b413e91d-40aa-a1a8-b8e3-2bacc3b3e222/00606949329020.rgb.jpg/100x100bb.jpg'
    };

    const { artistName, trackName, url } = data;

    const { getByTestId } = renderWithIntl(
      <ItuneCard artistName={artistName} trackName={trackName} artworkUrl100={url} />
    );

    expect(getByTestId('song-detail')).toHaveTextContent(artistName);
    expect(getByTestId('song-detail')).toHaveTextContent(trackName);
    expect(getByTestId('song-image')).toHaveAttribute('src', url);
  });
});
