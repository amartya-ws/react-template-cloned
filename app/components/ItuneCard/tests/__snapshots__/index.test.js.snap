// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItuneCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="itune-card"
    >
      <div
        class="ant-card ant-card-bordered ItuneCard__StyledCard-sc-1fg0mgb-0 lbzofK"
      >
        <div
          class="ant-card-cover"
        >
          <img
            alt="song cover"
            class="ItuneCard__StyledImg-sc-1fg0mgb-1 fpRYqQ"
            data-testid="song-image"
          />
        </div>
        <div
          class="ant-card-body"
        >
          <div
            class="ant-card-meta"
            data-testid="song-detail"
          >
            <div
              class="ant-card-meta-detail"
            >
              <div
                class="ant-card-meta-title"
              >
                not found
              </div>
              <div
                class="ant-card-meta-description"
              >
                not found
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
